src/Compiler/Checking/CheckDeclarations.fs

```diff
@@ -3349,12 +3349,26 @@ module EstablishTypeDefinitionCores =
 
             let multiCaseUnionStructCheck (unionCases: UnionCase list) =
                 if tycon.IsStructRecordOrUnionTycon && unionCases.Length > 1 then 
-                    let fieldNames = [ for uc in unionCases do for ft in uc.FieldTable.TrueInstanceFieldsAsList do yield (ft.LogicalName, ft.Range) ]
-                    let distFieldNames = fieldNames |> List.distinctBy fst
-                    if distFieldNames.Length <> fieldNames.Length then
-                        let fieldRanges = distFieldNames |> List.map snd
-                        for m in fieldRanges do
-                            errorR(Error(FSComp.SR.tcStructUnionMultiCaseDistinctFields(), m))
+                    let allTypesEqual listOfNamesAndTypes =
+                        match listOfNamesAndTypes with
+                        | [] | [_] -> true
+                        | (_,headField : RecdField) :: tail -> 
+                            let headType = headField.FormalType
+                            tail |> List.forall (fun (_,elemType) -> typeEquivAux EraseAll g elemType.FormalType headType)
+
+                    let diagnostics,duplicateCriteria =
+                        if cenv.g.langVersion.SupportsFeature(LanguageFeature.ReuseSameFieldsInStructUnions) then
+                            FSComp.SR.tcStructUnionMultiCaseFieldsSameType, allTypesEqual >> not
+                        else
+                            FSComp.SR.tcStructUnionMultiCaseDistinctFields, (fun group -> group.Length > 1)
+
+                    [ for uc in unionCases do for ft in uc.FieldTable.TrueInstanceFieldsAsList do yield(ft.LogicalName,ft) ]
+                    |> List.groupBy fst
+                    |> List.filter (fun (_name,group) -> duplicateCriteria group)                    
+                    |> List.iter (fun (_,dups) ->
+                        for _,ft in dups do
```

Comment:
The previous version actually had a reporting + range bug.
Instead of reporting all the duplicates, it reported all distinct occurrences of names.

Therefore:
Non-clashing names were also reported
Duplicate definitions were reported just once, all their other ranges were unreported.

While bringing in the new feature, it was more consistent to fix it here too.